File: backend/src/database/entities/asset.entity.ts
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryColumn,
  UpdateDateColumn,
} from 'typeorm';
import { sqliteDateTransformer } from '../../common/dates';
import { UserEntity } from './user.entity';

@Entity('asset_entities')
export class AssetEntity {
  @PrimaryColumn({ type: 'text' })
  id: string;

  @Column({ type: 'text', nullable: true })
  name: string;

  @Column({ type: 'text', nullable: true })
  symbol: string;

  @Column({ type: 'text', nullable: true })
  type: string;

  @Column({ type: 'real', nullable: true })
  quantity: number;

  @Column({ name: 'current_price', type: 'real', nullable: true })
  currentPrice: number;

  @Column({ name: 'user_id', type: 'integer', nullable: true })
  userId: number;

  @ManyToOne(() => UserEntity)
  @JoinColumn({ name: 'user_id' })
  user: UserEntity;

  @CreateDateColumn({ name: 'created_at', type: 'datetime', transformer: sqliteDateTransformer })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at', type: 'datetime', transformer: sqliteDateTransformer })
  updatedAt: Date;

  @DeleteDateColumn({ name: 'deleted_at', type: 'datetime', transformer: sqliteDateTransformer, nullable: true })
  deletedAt: Date | null;
}

File: backend/database/asset_entity.go
package database

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type AssetEntity struct {
	ID           uuid.UUID `gorm:"type:text;primaryKey"`
	Name         string
	Symbol       string
	Type         string
	Quantity     float64
	CurrentPrice float64

	UserID uint
	User   UserEntity

	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt gorm.DeletedAt `gorm:"index"`
}

type AssetTransactionEntity struct {
	ID uuid.UUID `gorm:"type:text;primaryKey"`
	Type        string
	Quantity    float64
	Price       float64
	Date        string
	Description string
	
	AssetID uuid.UUID
	Asset   AssetEntity

	UserID uint
	User   UserEntity

	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt gorm.DeletedAt `gorm:"index"`
}

func (l *AssetEntity) BeforeCreate(tx *gorm.DB) (err error) {
	if l.ID == uuid.Nil {
		l.ID = uuid.Must(uuid.NewV7())
	}
	return
}

func (l *AssetTransactionEntity) BeforeCreate(tx *gorm.DB) (err error) {
	if l.ID == uuid.Nil {
		l.ID = uuid.Must(uuid.NewV7())
	}
	return
}
